File: aula03/project/src/model/ordenacao.py
class Ordenacao:

    @staticmethod
    def bolha(lista):
        houve_troca = True
        qtd_comparacoes = 0
        qtd_trocas = 0
        fim = len(lista)

        while houve_troca:
            houve_troca = False
            for i in range(fim - 1):
                qtd_comparacoes += 1
                if lista[i] > lista[i + 1]:
                    qtd_trocas += 1
                    houve_troca = True
                    tmp = lista[i]
                    lista[i] = lista[i + 1]
                    lista[i + 1] = tmp
            fim -= 1

        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def selecao(lista):
        qtd_comparacoes = 0
        qtd_trocas = 0
        for i in range(len(lista) - 1):
            posMenor = i
            for j in range(i + 1, len(lista)):
                qtd_comparacoes += 1
                if lista[j] < lista[posMenor]:
                    posMenor = j
            if i != posMenor:
                qtd_trocas += 1
                tmp = lista[i]
                lista[i] = lista[posMenor]
                lista[posMenor] = tmp

        return qtd_comparacoes, qtd_trocas

    @staticmethod
    def insercao(lista):
        """
        Exemplo
        3, 1, 2
        3 > 1 → troca → 1, 3, 2
        3 > 2 → troca → 1, 2, 3
        """
        qtd_comparacoes = 0
        qtd_trocas = 0
        
        for i in range(1, len(lista)):
            j = i
            
            while j > 0:
                qtd_comparacoes += 1
                
                if lista[j-1] > lista[j]:
                #troca de lugar o indice, o que era o ultimo passa pra posicao do que ta olhando
                    lista[j-1], lista[j] = lista[j], lista[j-1]                
                    qtd_trocas += 1
                    j -= 1
                else:
                    break
        return qtd_comparacoes, qtd_trocas
  
     
    @staticmethod
    def sort(lista):
        lista.sort()
        return "N/A", "N/A"

File: aula07/paraContinuarEmLab/src/model/Ordenacao.java
package model;

import java.util.ArrayList;

/**
 * Algoritmos de ordenação usados no laboratório e suas métricas de desempenho.
 *
 * Cada método ordena a lista recebida e devolve duas métricas em
 * {ArrayList}: quantidade de comparações (índice 0) e quantidade de
 * trocas (índice 1).
 *
 */
public class Ordenacao {

    /**
     * Ordena a lista pelo método da bolha (Bubble Sort).
     *
     * <p>Percorre pares vizinhos e troca quando estão fora de ordem, repetindo
     * o processo enquanto houver troca em uma passagem completa.</p>
     *
     * @param lista Lista de inteiros que será ordenada in-place.
     * @return Métricas da execução: comparações no índice 0 e trocas no índice 1.
     */
    public static ArrayList bolha(ArrayList<Integer> lista) {
        ArrayList<Float> metricas = new ArrayList<>();
        long qtdComparacoes = 0;
        long qtdTrocas = 0;
        int aux;
        boolean houveTroca;
        int i;
        do {
            houveTroca = false;
            for (i = 0; i < lista.size() - 1; i++) {
                qtdComparacoes++;
                if (lista.get(i) > lista.get(i + 1)) {
                    houveTroca = true;
                    aux = lista.get(i);
                    lista.set(i, lista.get(i + 1));
                    lista.set(i + 1, aux);
                    qtdTrocas++;
                }
            }
        } while (houveTroca);
        metricas.add((float)qtdComparacoes);
        metricas.add((float)qtdTrocas);
        return metricas;
    }

    /**
     * Ordena a lista pelo método da seleção (Selection Sort).
     *
     * @param lista Lista de inteiros que será ordenada in-place.
     * @return Métricas da execução: comparações no índice 0 e trocas no índice 1.
     */
    public static ArrayList selecao(ArrayList<Integer> lista) {
        ArrayList<Float> metricas = new ArrayList<>();
        long qtdComparacoes = 0;
        long qtdTrocas = 0;
        int i, j, posMenor, aux;
        posMenor = 0;

        // A troca fica dentro do for (i). Fora do laço, i == lista.size()
        // e lista.get(i) gera IndexOutOfBoundsException.
        for (i = 0; i < lista.size(); i++) {
            posMenor = i;
            for (j = i+1; j < lista.size(); j++) {
                qtdComparacoes++;
                if (lista.get(j) < lista.get(posMenor)) {
                    posMenor = j;
                }
            }
            if (posMenor != i) {
                aux = lista.get(i);
                lista.set(i, lista.get(posMenor));
                lista.set(posMenor, aux);
                qtdTrocas++;
            }
        }
        metricas.add((float)qtdComparacoes);
        metricas.add((float)qtdTrocas);
        return metricas;
    }

    /**
     * Ordena a lista pelo método da inserção (Insertion Sort).
     *
     * @param lista Lista de inteiros que será ordenada in-place.
     * @return Métricas da execução: comparações no índice 0 e trocas no índice 1.
     */
    public static ArrayList insercao(ArrayList<Integer> lista) {
        ArrayList<Float> metricas = new ArrayList<>();
        long qtdComparacoes = 0;
        long qtdTrocas = 0;
        int i, j, aux;

        for (i = 1; i < lista.size(); i++) {
            aux = lista.get(i);
            for (j = i-1; j > 0 && aux < lista.get(j); j-- , qtdComparacoes++) {
                qtdTrocas++;
                lista.set(j+1, lista.get(j));
            }
            lista.set(j+1, aux);
        }
        metricas.add((float)qtdComparacoes);
        metricas.add((float)qtdTrocas);
        return metricas;
    }

    /**
     * Ordena a lista pelo método do pente (Comb Sort).
     *
     * @param lista Lista de inteiros que será ordenada in-place.
     * @return Métricas da execução: comparações no índice 0 e trocas no índice 1.
     */
    public static ArrayList pente(ArrayList<Integer> lista) {
        ArrayList<Float> metricas = new ArrayList<>();
        long qtdComparacoes = 0;
        long qtdTrocas = 0;
        int aux;
        boolean houveTroca;
        int i;
        int distancia = lista.size();
        do {
            distancia = (int)(distancia/1.3);
            if (distancia <= 0) {
                distancia = 1;
            }
            houveTroca = false;
            for (i = 0; i + distancia < lista.size(); i++) {
                qtdComparacoes++;
                if (lista.get(i) > lista.get(i + distancia)) {
                    houveTroca = true;
                    aux = lista.get(i);
                    lista.set(i, lista.get(i + distancia));
                    lista.set(i + distancia, aux);
                    qtdTrocas++;
                }
            }
        } while (distancia > 1 || houveTroca);
        metricas.add((float)qtdComparacoes);
        metricas.add((float)qtdTrocas);
        return metricas;
    }
}
